chore: remove implicit tokio feature

WaterWhisperer opened the pull request:
Fixes #511

<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/tower-rs/tower-http/blob/master/CONTRIBUTING.md
-->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? If a new feature is being added, describe the intended
use case that feature fulfills.
-->
> We have an implicit `tokio` Cargo feature because we don't use `dep:` syntax for it.

## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->
makes the `tokio` dependency explicit

WaterWhisperer committed:
chore: remove implicit tokio feature

Fixes #511

jplatte commented:
That CI error is really strange. Could you try moving the `rust-toolchain` action step in the MSRV job after the `cargo update` steps? Maybe this is an old bug in cargo that since got fixed.

WaterWhisperer commented:
> That CI error is really strange. Could you try moving the `rust-toolchain` action step in the MSRV job after the `cargo update` steps? Maybe this is an old bug in cargo that since got fixed.

Sure, thanks for the review

WaterWhisperer commented:
> Okay, so that didn't work. I have another idea: Can you (after rebasing, this PR seems to have a conflict now) move the toolchain action back, and also add `--all-features` to the problematic `cargo update` invocation? I think it's maybe saying that tokio is not a dependency that can be updated because it needs to resolve what dependencies actually exist and uses the default features for that unless told otherwise.

Seems that `cargo update` command does not support the `--all-features` flag.

WaterWhisperer commented:
This error is really strange :(

jlizen committed:
chore: use dep:tokio + tokio?/feat to work around MSRV resolver bug

The bare "tokio" feature in the previous commit triggers
rust-lang/cargo#10788 in our MSRV 1.64 CI job: `cargo update -p <crate>
--precise <ver>` fails because cargo can't resolve an optional dep
referenced only via `dep:` syntax. Using `tokio?/<subfeature>` keeps
the dep name visible to the resolver while still avoiding the implicit
feature. Same pattern tottoto used in #642 for async-compression.

Also adds a CHANGELOG entry for the breaking change.

jlizen commented:
Updated to use @tottoto's approach, thanks:

```
  -metrics = ["dep:http-body", "tokio/time"]
  +metrics = ["dep:http-body", "dep:tokio", "tokio?/time"]
```
